apps/meteor/app/models/server/raw/Sessions.ts

```diff
@@ -737,12 +739,333 @@ export const aggregates = {
 };
 
 export class SessionsRaw extends BaseRaw<ISession> {
+	protected indexes: IndexSpecification[] = [
+		{ key: { instanceId: 1, sessionId: 1, year: 1, month: 1, day: 1 } },
+		{ key: { instanceId: 1, sessionId: 1, userId: 1 } },
+		{ key: { instanceId: 1, sessionId: 1 } },
+		{ key: { sessionId: 1 } },
+		{ key: { userId: 1 } },
+		{ key: { userId: 1, sessionId: 1 } },
+		{ key: { userId: 1, loginToken: 1, logoutAt: 1 } },
+		{ key: { loginToken: 1, logoutAt: 1 } },
+		{ key: { loginToken: 1 } },
+		{ key: { year: 1, month: 1, day: 1, type: 1 } },
+		{ key: { type: 1 } },
+		{ key: { ip: 1, loginAt: 1 } },
+		{ key: { createdAt: 1 } },
+		{ key: { _computedAt: 1 }, expireAfterSeconds: 60 * 60 * 24 * 45 },
+		{
+			key: {
+				'device.name': 'text',
+				'device.type': 'text',
+				'device.os.name': 'text',
+				'sessionId': 'text',
+				'userId': 'text',
+			},
+			weights: {
+				'device.name': 10,
+				'device.os.name': 10,
+				'device.type': 5,
+				'sessionId': 1,
+				'userId': 1,
+			},
+			name: 'text_search_w_weights',
+		},
+	];
+
 	private secondaryCollection: Collection<ISession>;
 
 	constructor(public readonly col: Collection<ISession>, public readonly colSecondary: Collection<ISession>, trash?: Collection<ISession>) {
 		super(col, trash);
 
 		this.secondaryCollection = colSecondary;
+		this.col.createIndexes(this.indexes);
+	}
+
+	async getByUserId(
+		uid: string,
+		search?: string | null,
+		{ offset, count }: IPaginationOptions = { offset: 0, count: 50 },
+	): Promise<PaginatedResult<{ sessions: ISession[] }>> {
+		const searchQuery = search ? [{ $text: { $search: search } }] : [];
+
+		const matchOperator = {
+			$match: {
+				$and: [
+					...searchQuery,
+					{
+						userId: {
+							$eq: uid,
+						},
+					},
+					{
+						loginToken: {
+							$exists: true,
+							$ne: '',
+						},
+					},
+					{
+						logoutAt: {
+							$exists: false,
+						},
+					},
+				],
+			},
+		};
+		const sortOperator = {
+			$sort: {
+				createdAt: -1,
+			},
+		};
+		const groupOperator = {
+			$group: {
+				_id: '$loginToken',
+				instanceId: {
+					$first: '$instanceId',
+				},
+				sessionId: {
+					$first: '$sessionId',
+				},
+				day: {
+					$first: '$day',
+				},
+				month: {
+					$first: '$month',
+				},
+				year: {
+					$first: '$year',
+				},
+				userId: {
+					$first: '$userId',
+				},
+				device: {
+					$first: '$device',
+				},
+				host: {
+					$first: '$host',
+				},
+				ip: {
+					$first: '$ip',
+				},
+				mostImportantRole: {
+					$first: '$mostImportantRole',
+				},
+				roles: {
+					$first: '$roles',
+				},
+				type: {
+					$first: '$type',
+				},
+				_updatedAt: {
+					$first: '$_updatedAt',
+				},
+				createdAt: {
+					$first: '$createdAt',
+				},
+				loginAt: {
+					$first: '$loginAt',
+				},
+			},
+		};
+
+		const [docTotal] = await this.col
```

Comment:
Instead of running the agg 2 times, you can take a look at the `$facet` stage 👀 
https://www.mongodb.com/docs/manual/reference/operator/aggregation/facet/

(If you search for $facet on the repo, you'll find 2 examples of this being used to paginate aggregations on a single call 👀 )

Reply:
So nice!
We have many things that follow this pattern above. Maybe updating this pattern is a good task that we can do in a near future. :smiley: